File: auth-service/app/oauth/discord.py
import httpx
from app.config import settings

async def get_discord_token(code: str):
    async with httpx.AsyncClient() as client:
        response = await client.post("https://discord.com/api/oauth2/token", data={
            "client_id": settings.DISCORD_CLIENT_ID,
            "client_secret": settings.DISCORD_CLIENT_SECRET,
            "grant_type": "authorization_code",
            "code": code,
            "redirect_uri": settings.DISCORD_REDIRECT_URI
        }, headers={"Content-Type": "application/x-www-form-urlencoded"})
        return response.json()

async def get_discord_user(token: str):
    async with httpx.AsyncClient() as client:
        response = await client.get("https://discord.com/api/users/@me",
                                    headers={"Authorization": f"Bearer {token}"})
        return response.json()

File: auth-service-ts/src/oauth/discord.ts
import { config } from '../config';

const DISCORD_OAUTH_URL = 'https://discord.com/api/oauth2/authorize';
const DISCORD_TOKEN_URL = 'https://discord.com/api/oauth2/token';
const DISCORD_USER_URL = 'https://discord.com/api/users/@me';

export function getDiscordOAuthUrl(): string {
  const params = new URLSearchParams({
    client_id: config.discord.clientId,
    redirect_uri: config.discord.redirectUri,
    response_type: 'code',
    scope: 'identify email',
    prompt: 'consent',
  });
  return `${DISCORD_OAUTH_URL}?${params.toString()}`;
}

export async function exchangeDiscordCodeForToken(code: string) {
  const params = new URLSearchParams({
    client_id: config.discord.clientId,
    client_secret: config.discord.clientSecret,
    grant_type: 'authorization_code',
    code,
    redirect_uri: config.discord.redirectUri,
  });
  const resp = await fetch(DISCORD_TOKEN_URL, {
    method: 'POST',
    headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
    body: params.toString(),
  });
  if (!resp.ok) throw new Error('Failed to exchange code for token');
  return await resp.json(); // { access_token, token_type, ... }
}

export async function fetchDiscordUser(accessToken: string) {
  const resp = await fetch(DISCORD_USER_URL, {
    headers: { 'Authorization': `Bearer ${accessToken}` },
  });
  if (!resp.ok) throw new Error('Failed to fetch Discord user');
  return await resp.json(); // { id, username, email, ... }
}
